### ISSUE TITLE
File names containing backslashes can shadow files in subdirectories on Linux

### ISSUE THREAD

**Issue by cameel** [trufflesuite/truffle#4166]
## Issue
On Linux backslashes (`\`) can be freely used in file names and are valid characters. Truffle, however, normalizes them into forward slashes. When both a name with a backslash (`a\b.sol`) and a file in a subdirectory (`a/b.sol`) are present, Truffle chooses the former and silently ignores the latter.

I think that Truffle should either include both files in the input or report an error when it encounters such a situation.

## Steps to Reproduce
```bash
mkdir truffle-test/
cd truffle-test/
npm install truffle
npx truffle init
echo 'module.exports.compilers.solc.version = "0.8.6";' >> truffle-config.js

mkdir contracts/c/
echo 'contract C1 {}' > contracts/c/C.sol
echo 'contract C2 {}' > contracts/c\\C.sol
npx truffle compile
```

## Actual Results
```
> Artifacts written to /tmp/truffle-test/build/contracts
> Compiled successfully using:
   - solc: 0.8.6+commit.11564f7e.Emscripten.clang
```

The `sources` dict in the JSON input contains these entries:
```json
"project:/contracts/c/C.sol": { "content": "contract C2 {}\n" }
```

## Expected Behavior
Both contracts should be available and stored under the correct names:
```json
"project:/contracts/c/C.sol": { "content": "contract C1 {}\n" },
"project:/contracts/c\\C.sol": { "content": "contract C2 {}\n" }
```

## Environment
* Operating System: Arch Linux
* Truffle version: both `5.3.14` and current `develop` (abfb3646eea9d57e0a6f9186c12bb51680c181d8)
* node version: `v16.4.2`
* npm version: `7.17.0`

**Comment by haltman-at** [comment]
Hm, OK, this shouldn't be too hard a change -- unles @gnidan is right and `path.sep` isn't always reliable...?

**Comment by eggplantzzz** [comment]
Thanks @cameel, we appreciate you taking the time to write these issues and help us improve! This should be resolved with Truffle v5.4.1.
